### ISSUE TITLE
bug with OPTIONAL MATCH and DISTINCT

### ISSUE THREAD

**Issue by matthew-s-miller** [neo4j/neo4j#12232]
I've discovered a query that uses OPTIONAL MATCH and DISTINCT for which cypher returns the wrong result

**Neo4j Version:** 3.5.6
**Operating System:** Windows 10
**API:** Cypher

### Steps to reproduce
1. Create the following data (note the different `:Bar`, `:Baz` labels used)
```
CREATE (f:Foo {uuid: 1})
CREATE (f)-[:BAR]->(:Bar)-[:BEL]->(:Bel)
CREATE (f)-[:BAR]->(:Baz)-[:BEL]->(:Bel)
```
2. Run this query:
```
OPTIONAL MATCH (foo:Foo {uuid: 1})-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel)
RETURN COUNT(DISTINCT bel)
```

### Expected behavior
The result should be a count of **1**

### Actual behavior
 a count of **2** is returned

### Synopsis
The following versions of the query all return the correct result...
```
MATCH (foo:Foo {uuid: 1})-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel) RETURN COUNT(DISTINCT bel)
```
```
OPTIONAL MATCH (foo:Foo {uuid: 1})-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel) RETURN COUNT(bel)
```
```
OPTIONAL MATCH (foo:Foo)-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel) RETURN COUNT(DISTINCT bel)
```

After profiling the bad query, it seems that neo4j is omitting **Filter** steps from its execution plan

**Comment by InverseFalcon** [comment]
Able to replicate this, and yes it looks like expected label filtering isn't happening.

If we do a MATCH first to `foo` then an OPTIONAL MATCH for the rest then we seem to get the right count:

```
MATCH (foo:Foo {uuid: 1}) 
OPTIONAL MATCH (foo)-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel)
RETURN COUNT(DISTINCT bel)
```

That seems to work whether there's an index present on :Foo(uuid) or not.

This can be used as a workaround for the time being.

**Comment by matthew-s-miller** [comment]
Thanks for that. Interestingly you can also start with an OPTIONAL MATCH too!
```
OPTIONAL MATCH (foo:Foo {uuid: 1})
OPTIONAL MATCH (foo)-[:BAR]->(bar:Bar)-[:BEL]->(bel:Bel)
RETURN COUNT(DISTINCT bel)
```

**Comment by matthew-s-miller** [comment]
Another interesting observation is that in the failing query the planner chooses to start at `Bel` and work backwards. Whereas the successful queries all start with `Foo`.

Our workaround queries essential enforce that the planner starts at `Foo` also.

**Comment by Lojjs** [comment]
@matthew-s-miller It turned out to be a bug in the rewriter step of optional match which only caused problems with distinct and when the expand start from the `:Bel` node, which your alternative queries don't fulfill. As you noticed, the filter of the label predicates were dropped, which is wrong. I have now merged a fix for the bug which just made 3.5.7, which will be released within the next few days.

Best regards Louise, Neo4j Cypher team
